Add FeesDifferentThanExpected test to TransactionFactoryTests

yahiheb opened the pull request:
This test simulates the `InvalidTxException: Invalid transaction` As suggested here https://github.com/zkSNACKs/WalletWasabi/issues/6750#issuecomment-988917870

kiminuo commented:
The test looks very good. 

Some nits: https://github.com/kiminuo/WalletWasabi/tree/different-feerate-test

Could you please describe the user scenario a bit more? [Here](https://github.com/zkSNACKs/WalletWasabi/issues/6750#issue-1063745395) you mention: 

> Basically to reproduce this issue one needs to try to spend one whole coin without getting any change: `the whole coin = the sent amount + the fee`

But I'm still unclear whether this is a valid user scenario or not. I mean a user may want to spend some coin without wanting to have a change coin and in that case WW may report that "WalletWasabi.Exceptions.InvalidTxException: Invalid transaction", right? (I don't know GUI enough.)

Is that correct description of the issue?

yahiheb commented:
> I mean a user may want to spend some coin without wanting to have a change coin and in that case WW may report that "WalletWasabi.Exceptions.InvalidTxException: Invalid transaction", right?

No not always.

If I am correct the exception happens when we build a transaction that "generate" change but it is uneconomical, so that change won't be created and it gets added to the fees which results in this difference of fees between the expected fee and the actual one.

nopara73 reviewed (dismissed):
LGTM

kiminuo commented:
@yahiheb I would say it would be nice to improve the test like this: https://github.com/kiminuo/WalletWasabi/commit/330ef6f1e57541c921125b2257a66d304c3bf452, merge this PR and then modify the test in #6909 and merge that PR. 

This would nicely "document" the changes being done and that indeed the number of output coins decreases from 2 to 1 as expected.

yahiheb commented:
> @yahiheb I would say it would be nice to improve the test like this: [kiminuo@330ef6f](https://github.com/kiminuo/WalletWasabi/commit/330ef6f1e57541c921125b2257a66d304c3bf452), merge this PR and then modify the test in #6909 and merge that PR.
> 
> This would nicely "document" the changes being done and that indeed the number of output coins decreases from 2 to 1 as expected.

@kiminuo I added your commit, and refactored the code a bit https://github.com/zkSNACKs/WalletWasabi/pull/6871/commits/db5749b7c5ac45221692ecc23a0a306adf617606, thank you.

kiminuo reviewed (approved):
utACK

No big risk in merging. It's just a test.
